mitosheet/mitosheet/public/v3/sheet_functions/string_functions.py

```diff
@@ -8,18 +8,79 @@
 
 NOTE: This file is alphabetical order!
 """
-from typing import Union
+from typing import Optional, Union
 import sys
 import numpy as np
 
 import pandas as pd
 
+from mitosheet.public.v3.errors import handle_sheet_function_errors
 from mitosheet.public.v3.sheet_functions.utils import get_final_result_series_or_primitive
-from mitosheet.public.v3.types.decorators import cast_values_in_arg_to_type
-from mitosheet.public.v3.types.sheet_function_types import StringFunctionReturnType
+from mitosheet.public.v3.types.decorators import cast_values_in_args_to_type, cast_values_in_arg_to_type
+from mitosheet.public.v3.types.sheet_function_types import StringInputType, StringRestrictedInputType, StringFunctionReturnType, IntRestrictedInputType
 
-@cast_values_in_arg_to_type('str')
-def CONCAT(*argv: pd.Series) -> StringFunctionReturnType:
+
+@cast_values_in_arg_to_type('string', 'str')
+@cast_values_in_arg_to_type('num_chars', 'int')
+@handle_sheet_function_errors
+def LEFT(string: StringRestrictedInputType, num_chars: IntRestrictedInputType=None) -> StringFunctionReturnType:
+    """
+    {
+        "function": "LEFT",
+        "description": "Returns a substring from the beginning of a specified string.",
+        "search_terms": ["left"],
+        "examples": [
+            "LEFT(A, 2)",
+            "LEFT('The first character!')"
+        ],
+        "syntax": "LEFT(string, [number_of_characters])",
+        "syntax_elements": [{
+                "element": "string",
+                "description": "The string or series from which the left portion will be returned."
+            },
+            {
+                "element": "number_of_characters [OPTIONAL, 1 by default]",
+                "description": "The number of characters to return from the start of string."
+            }
+        ]
+    }
+    """
+
+    if string is None:
+        return ''
+    else:
+
+        if num_chars is None:
+            num_chars = 1
+        elif isinstance(num_chars, pd.Series):
+            num_chars = num_chars.fillna(1)
+
+        if isinstance(string, str):
+            if isinstance(num_chars, int):
+                return string[:num_chars]
+            else:
+                print(num_chars)
+                return pd.Series(
+                    [string[:nc] for nc in num_chars],
+                    index=num_chars.index
+                )
+        else:
+            if isinstance(num_chars, int):
+                print("HERE!", num_chars, [s[:num_chars] for s in string])
+                return pd.Series(
+                    [s[:min(num_chars, len(s))] for s in string],
+                    index=string.index
+                )
+            else:
+                return pd.Series(
+                    [s[:nc] for s, nc in zip(string, num_chars)],
```

Comment:
TODO: Make sure there is a text here where the num_chars in the series is longer than the string, since we are not doing the min check